libp2p/crypto/crypto.nim

```diff
@@ -79,146 +82,165 @@ type
   Signature* = object
     data*: seq[byte]
 
-  P2pKeyError* = object of CatchableError
-  P2pSigError* = object of CatchableError
+  CryptoError* = enum
+    KeyError,
+    SigError,
+    HashError,
+    SchemeError
+
+  CryptoResult*[T] = Result[T, CryptoError]
 
 const
   SupportedSchemes* = {RSA, Ed25519, Secp256k1, ECDSA}
   SupportedSchemesInt* = {int8(RSA), int8(Ed25519), int8(Secp256k1),
                           int8(ECDSA)}
 
 proc random*(t: typedesc[PrivateKey], scheme: PKScheme,
-             bits = DefaultKeySize): PrivateKey =
+             bits = DefaultKeySize): CryptoResult[PrivateKey] =
   ## Generate random private key for scheme ``scheme``.
   ##
   ## ``bits`` is number of bits for RSA key, ``bits`` value must be in
   ## [512, 4096], default value is 2048 bits.
-  doAssert(scheme in SupportedSchemes)
-  result = PrivateKey(scheme: scheme)
-  if scheme == RSA:
-    result.rsakey = RsaPrivateKey.random(bits)
-  elif scheme == Ed25519:
-    result.edkey = EdPrivateKey.random()
-  elif scheme == ECDSA:
-    result.eckey = EcPrivateKey.random(Secp256r1)
-  elif scheme == Secp256k1:
-    result.skkey = SkPrivateKey.random()
+  if scheme notin SupportedSchemes:
+    err(SchemeError)
+  else:
+    var res: t
+    res = PrivateKey(scheme: scheme)
+    if scheme == RSA:
+      res.rsakey = ? (RsaPrivateKey.random(bits).mapErr do (_: auto) -> auto: KeyError)
```

Comment:
Hmmm had a quick go but
```
cannot prove that it's safe to initialize 'rsakey' with the runtime value for the discriminator 'scheme'
```

Reply:
just use `RSA` like in the comparison? repetitive unfortunately, but..

also, this whole construct should use `case scheme of` - it could then perhaps become error-free (since all case enums must be checked)